src/adapters/thinkerbell/mod.rs

```diff
@@ -0,0 +1,404 @@
+                                    error!("Unable to remove Thinkerbell Rule Service: {:?}", e)
+                                }
+                            }
+                            break;
+                        }
+                    }
+                },
+                // Respond to a pending Getter request.
+                ThinkAction::RespondToGetter(tx, getter_id) => {
+                    for ref rule in &rules {
+                        if getter_id == rule.getter_is_enabled_id {
+                            let is_enabled = script_manager.is_enabled(&rule.script_id);
+                            let _ = tx.send(Ok(Some(Value::OnOff(if is_enabled { OnOff::On } else { OnOff::Off }))));
+                            continue 'recv;
+                        } else if getter_id == rule.getter_source_id {
+                            match script_manager.get_source(&rule.script_id) {
+                                Ok(source) => {
+                                    let _ = tx.send(Ok(Some(Value::String(Arc::new(source.to_owned())))));
+                                },
+                                Err(e) => {
+                                    let _ = tx.send(Err(sm_error(e)));
+                                }
+                            };
+                            continue 'recv;
+                        }
+                    }
+                    let _ = tx.send(Err(Error::InternalError(InternalError::NoSuchGetter(getter_id.clone()))));
+                },
+                // Respond to a pending Setter request.
+                ThinkAction::RespondToSetter(tx, setter_id, value) => {
+                    // Add a new rule (with the given JSON source).
+                    if setter_id == self.setter_add_rule_id {
+                        match value {
+                            Value::ThinkerbellRule(ref rule_source) => {
+                                let script_id = Id::new(&rule_source.name);
+                                let _ = tx.send(script_manager.put(&script_id, &rule_source.source).map_err(sm_error));
+                                let _ = self.tx.send(ThinkAction::AddRuleService(script_id.clone()));
+                            },
+                            _ => {
+                                let _ = tx.send(Err(Error::TypeError(TypeError {
+                                    expected: Type::ThinkerbellRule,
+                                    got: value.get_type()
+                                })));
+                            }
+                        }
+                    } else {
+                        // The rest of the rules are script/rule-specific.
+                        // NOTE: This linear search is not ideal, but tracking getters/setters in maps
+                        // would be far more complex until we have a simpler way to track state within
+                        // getter/setter API requests. In any case, this loop should be plenty fast for now.
```

Comment:
If you have an API suggestion, please file it. Now that everybody has tested the API, it's the right time to improve it before we proceed.